kind-ovn: Introduce new kind-ovn provider

oshoval opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Provider allows running kubevirt with ovn-k8s tests on CI.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

oshoval commented:
/test check-provision-manager

oshoval commented on cluster-up/cluster/kind-ovn/provider.sh:
```diff
@@ -0,0 +1,92 @@
+    cluster::install
+    fetch_kind
+
+    if [[ ${CI} == "true" ]]; then
+        sysctl fs.inotify.max_user_watches=1048576
+        sysctl fs.inotify.max_user_instances=512
+        dnf install -y openssl
+    fi
```
will be removed once we can use the updated legacy image
https://github.com/kubevirt/project-infra/pull/3431

oshoval commented:
Thanks for the help here @qinqon

qinqon commented on cluster-up/cluster/kind-ovn/provider.sh:
```diff
@@ -0,0 +1,91 @@
+KIND_PATH=${KIND_PATH:-"${KUBEVIRTCI_CONFIG_PATH}/${KUBEVIRT_PROVIDER}/_kind"}
+CLUSTER_PATH=${CLUSTER_PATH:-"${KUBEVIRTCI_CONFIG_PATH}/${KUBEVIRT_PROVIDER}/_ovnk"}
+
+MTU=${MTU:-"1392"}
```
Let's try to calculate it taking the MTU of the interface from the default gw route

```bash
overlay_overhead=58
current_mtu=$(cat /sys/class/net/$(ip route | grep "default via" | head -1 | awk '{print $5}')/mtu)
overlay_mtu=$(expr $current_mtu - $overlay_overhead)
```

oshoval replied:
but why ? there is no need, one can just export MTU in case he has smaller MTU
and it is rare
we can just add readme instead
and make the code simpler

maiqueb replied:
If something changes on the prow infra, your code will break, you won't know why, you'll have to debug, spend time on it, eventually figure out the MTU changed, and update one magic value by another magic value.

Quique's suggestion will work under those circumstances. 

> it is rare

? How so ? Why do you think this is strange ?

qinqon replied:
> but why ? there is no need, one can just export MTU in case he has smaller MTU and it is rare we can just add readme instead and make the code simpler

@oshoval by default this should try to calculate it with something like the bash snippet and also give users a scape hatch in case the calculations are wrong.

oshoval replied:
> ? How so ? Why do you think this is strange ?

based on the machines i saw, usually MTU is not that small
anyhow i will do that

qinqon commented on cluster-up/cluster/kind-ovn/provider.sh:
```diff
@@ -0,0 +1,91 @@
+    cluster::install
+    fetch_kind
+
+    if [[ ${CI} == "true" ]]; then
+        sysctl fs.inotify.max_user_watches=1048576
+        sysctl fs.inotify.max_user_instances=512
+        dnf install -y openssl
+    fi
```
I remember needing this at my laptop too, not only CI.

oshoval replied:
user would just need to do it once manually imo, we can add a readme
(locally i didnt but this is just a specific machine)

maiqueb replied:
Configuring the max user instances is something I have to do locally (I use podman).
Not sure about the max user watches.

qinqon replied:
> user would just need to do it once manually imo, we can add a readme, wdyt ? (locally i didnt but this is just a specific machine)
> 
> btw maybe worth without the watches on CI ? or you saw you must have them on CI ?

This is idempotent configuration, not big deal with doing it every time, let's just remove the CI env var and that's it.

oshoval replied:
note this is not something we are doing on any other provider,
i mean if some operations like installing stuff, config (for example disable_ipv6=0) etc
we require the user to do it just once, i think it is wrong and might require root (maybe we need root anyhow)
but if you insist i will do that
note that soon we will have a CI image that have 3 lines already, so it will be required only locally
which is even worse

maiqueb replied:
I really don't follow your argument @oshoval

oshoval replied:
ok tested with the new CI image `quay.io/kubevirtci/golang-legacy:v20240523-1a2c9b8`
we don't need on CI those lines anymore because now they are part of the image
```
        sysctl fs.inotify.max_user_watches=1048576
        sysctl fs.inotify.max_user_instances=512
        dnf install -y openssl
```
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/3426/rehearsal-pull-kubevirt-e2e-net-integration-tests/1793555893502087168

so now we don't need those on CI, lets talk about locally
until today, all the providers we created / have, don't install packages that are needed on the root level
(referring to openssl that was required), it is like if the provider need `wget` it won't install, the user need to.
and they also don't set configuration, for example disable_ipv6 must be 0, and it is the user responsibility to set it, the provider doesn't do it it each time

So please reconsider, do you still want to include those lines that now only required locally,
or to just add a readme that say what is needed (or even just add it on the PR desc) ?
fwiw we don't get into resolution in the readme of what should be dnf installed, and even about some of the configuration (such as disable_ipv6=0) we don't usually.

maiqueb replied:
I really don't care. 

As I see it, the 2 options are:
- hand a command to the user that just works. downside is ?... I don't see one.
- hand a command to the user that depends on the configuration of their environment. If / once the commands fails on their environment, the user will need to understand what's happening / look for a readme, and understand the dependencies, then act upon them

If the kubevirtci way is to do the latter, by all means, apply that. I know what I would rather have as a user.

qinqon commented on cluster-up/cluster/kind-ovn/provider.sh:
```diff
@@ -0,0 +1,91 @@
+        dnf install -y openssl
+    fi
+
+    pushd $CLUSTER_PATH/contrib ; ./kind.sh --multi-network-enable -mtu $MTU -lr --enable-interconnect; popd
```
We should also specify a different `--cluster-name` so it do not collide with the default ovn-k cluster, maybe `--cluster-name=kubevirtci`

maiqueb replied:
Good point; I'd go with the lane's name though.

qinqon commented on cluster-up/cluster/kind-ovn/provider.sh:
```diff
@@ -0,0 +1,91 @@
+        dnf install -y openssl
+    fi
+
+    pushd $CLUSTER_PATH/contrib ; ./kind.sh --multi-network-enable -mtu $MTU -lr --enable-interconnect; popd
```
Let's also use descriptive names here, so instead of `-lr` do `--local-registry`

qinqon commented on cluster-up/cluster/kind-ovn/provider.sh:
```diff
@@ -0,0 +1,91 @@
+}
+
+function down() {
+    echo "Not supported yet by this provider"
```
The implementation of this is `kind delete cluster --name $CLUSTER_NAME`

oshoval commented:
trying something, please dont review it yet
trying the new image quay.io/kubevirtci/golang-legacy:v20240523-1a2c9b8
that should have the 3 lines (watches / opensssl) built in

oshoval commented:
ready to review again please
addressed comments

using updated legacy image that set watches / install openssl already
bumped OVN k8s to include the e2e tests that were merged
testing the changes on CI for kubevirtci and kubevirt
https://github.com/kubevirt/project-infra/pull/3426
https://github.com/kubevirt/project-infra/pull/3430

note that on kubevirt we don't need `export DOCKER_PREFIX=localhost:5000` anymore,
as the provider writes the config file that kubevirt uses

oshoval commented:
rebased, and amend commit msg

brianmcarey reviewed (commented):
@oshoval can you add a short README with the steps for running it locally just so that we have it captured?

oshoval commented on cluster-up/cluster/kind-ovn/provider.sh:
```diff
@@ -0,0 +1,91 @@
+function calculate_mtu() {
+    overlay_overhead=58
+    current_mtu=$(cat /sys/class/net/$(ip route | grep "default via" | head -1 | awk '{print $5}')/mtu)
+    echo $(expr $current_mtu - $overlay_overhead)
```
we do need, else it tried to exec the value and fails
...provider.sh: line 30: 1442: command not found

oshoval commented:
addressed comments of @qinqon @brianmcarey 
thx

brianmcarey reviewed (commented):
Looks good to me - just one question inline.

oshoval commented on cluster-up/cluster/kind-ovn/provider.sh:
```diff
@@ -0,0 +1,91 @@
+x86_64* | i?86_64* | amd64*)
+    ARCH="amd64"
+    ;;
+ppc64le)
```
just carry over, will remove

oshoval commented:
addressed comments

maiqueb commented:
/lgtm

brianmcarey reviewed (approved):
/approve
